lib/cog/command/user_permissions_cache.ex

```diff
@@ -83,16 +83,38 @@ defmodule Cog.Command.UserPermissionsCache do
 
   defp fetch_and_cache({adapter, username}=key, state) do
     Logger.info("Cache miss for #{inspect key}")
-    case Repo.one(Queries.User.for_handle(username, adapter)) do
+
+    with {:ok, external_user} <- fetch_user_from_adapter(adapter, username),
+         {:ok, internal_user} <- fetch_user_from_database(adapter, external_user.id) do
+      perms = Cog.Models.User.all_permissions(internal_user)
+      value = {internal_user, perms}
+      expiry = Cog.Time.now() + state.ttl
+      :ets.insert(@ets_table, {key, value, expiry})
+      {:ok, value}
+    end
+  end
+
+  def fetch_user_from_adapter(adapter, handle) do
+    user = with {:ok, adapter_module} <- Cog.adapter_module(adapter),
+                do: adapter_module.lookup_user(handle: handle)
+
+    case user do
+      {:ok, user} ->
+        {:ok, user}
+      {:error, _} ->
+        {:error, :not_found}
+    end
+  end
```

Comment:
Also I guess with the `with` version, you can quickly look to see what the possible return values would be: `{:ok, user}` or `{:error, :not_found}`.